bigquery/google/cloud/bigquery/_helpers.py

```diff
@@ -122,6 +123,36 @@ def _record_from_json(value, field):
 }
 
 
+def _row_from_json(row, schema):
+    """Convert JSON row data to row with appropriate types.
+
+    :type row: dict
+    :param row: A JSON response row to be converted.
+
+    :type schema: tuple
+    :param schema: A tuple of
+                   :class:`~google.cloud.bigquery.schema.SchemaField`.
+
+    :rtype: tuple
+    :returns: A tuple of data converted to native types.
+    """
+    row_data = []
+    for field, cell in zip(schema, row['f']):
```

Comment:
That would be guarding against either a) a back-end bug (server returns rows which don't match the table's schema), or b) the user manually misconfiguring the schema on the client side, or failing to set it / load it all all (in which case they would get back "short" or empty record tuples).  I can see guarding against the "schema never set" case, but in the caller to this helper (`Table.fetch_data`) rather than here.